onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder.cc

```diff
@@ -1608,9 +1609,23 @@ bool GemmOpBuilder::IsQuantizedOp(const NodeUnit& node_unit) const {
 void GemmOpBuilder::AddInitializersToSkip(ModelBuilder& model_builder, const NodeUnit& node_unit) const {
   const auto& inputs = node_unit.Inputs();
   if (IsQuantizedOp(node_unit)) {
-    AddQuantizationScaleAndZeroPointToSkip(model_builder, *inputs[0].quant_param);               // b_scale, b_zp
-    AddInputToSkip(model_builder, inputs[1]);                                                    // b, b_scale, b_zp
-    AddQuantizationScaleAndZeroPointToSkip(model_builder, *node_unit.Outputs()[0].quant_param);  // y_scale, y_zp
+    if (node_unit.OpType() == "QlinearMatMul") {
+      AddQuantizationScaleAndZeroPointToSkip(model_builder, *inputs[0].quant_param);               // a_scale, a_zp
+      AddInputToSkip(model_builder, inputs[1]);                                                    // b, b_scale, b_zp
+      AddQuantizationScaleAndZeroPointToSkip(model_builder, *node_unit.Outputs()[0].quant_param);  // y_scale, y_zp
+    }
+    if (node_unit.OpType() == "Gemm") {                                               // QDQGemm
+      AddQuantizationScaleAndZeroPointToSkip(model_builder, *inputs[0].quant_param);  // x_scale, x_zp
+      AddQuantizationScaleAndZeroPointToSkip(model_builder, *inputs[1].quant_param);  // w_scale, w_zp
+      NodeAttrHelper helper(node_unit);
+      const auto transB = helper.Get("transB", 0);
+      if (transB == 0)
```

Comment:
Even transB is 1, you still need to skip the input if it's an initializer
So for this one you may want to check if inputs[1].node_arg is an initializer instead of checking if transB is 0

Reply:
Does the latest revision still need this check added or is it fine to blindly call AddInputToSkip? Maybe there's no negative consequence of adding a non-initializer to the list of things to skip.

Reply:
The original check is for skipping the second input (inputs[1].node_arg.Name()) , and I guess here AddInputToSkip()  deals with the bias input (inputs[2]) and skipping all of them here is similar to what we did for the qdq conv/qlinearconv operator where bias needs special handling. 

And removing the original check for skipping the initializer of the second input here because I found out "an operand indices key not found exception", so figured here maybe the initializer for input B can not be skipped. And the test passes with this.

Reply:
Sorry not quite understanding this. 

Don't we skip adding an initializer if we add it to the NNAPI model elsewhere (e.g. via AddInitializerTransposed) due to it needing some processing from the original value in the ONNX model to what is required in the NNAPI model?

That seems orthogonal to an error about a missing indices which I would have come from a test where the optional bias input was not provided.

Reply:
I am sorry yes. sorry what I meant if we skip by the condition when it is an initializer, then it will cause error. as for the transB=1 case, we actually didn't have additional handling to embed it into nnapi model.

so adding back the original check to skip the initializer on transB = 0 won't cause exception...

Reply:
I understand adding the initializer to skip if transB == 0, as we transpose that initializer separately and add it to the nnapi model as part of that process. 

If transB == 1 and that input is an initializer we can use it as is, and I assume the default handling will copy that initializer into the nnapi model.

Not following AddInputToSkip on input[2], which is the 'c' input isn't it and not the 'b' as the comment says? 

I don't see any requirement in GemmOpSupportChecker::IsOpSupportedImpl for the 'c' input to be an initializer. Did I miss where that happens? It looked like we just checked the shape to make sure it broadcasts correctly, and use it as is. 

Given that, why do we need to call AddInputToSkip, and why do we not treat it the same way as the 'a' input and just call AddQuantizationScaleAndZeroPointToSkip for it?